File: src/surfaces/contextInspector.ts
import { newId } from "../ids.js";
import { FabricError } from "../runtime/errors.js";
import { asRecord, getArray, getField, getOptionalString, getString, safeJsonArray, safeJsonRecord } from "../runtime/input.js";
import type { CallContext } from "../types.js";
import type { SurfaceHost } from "./host.js";

type ContextPackageInput = {
  preflightId: string;
  sessionId: string;
  agentId: string;
  hostName: string | null;
  workspaceRoot: string;
  client: string;
  taskType: string;
  inputTokens: number;
  contextPackage: Record<string, unknown>;
  contextSummary: Record<string, unknown>;
  toolSchemas: unknown[];
  mcpServers: unknown[];
  memories: unknown[];
  sensitiveFlags: string[];
  originPeerId: string;
  traceId?: string;
  correlationId?: string;
  testMode: boolean;
};

type ContextPackageRow = {
  id: string;
  preflight_request_id: string;
  ts: string;
  session_id: string;
  agent_id: string;
  host_name: string | null;
  workspace_root: string;
  client: string;
  task_type: string;
  input_tokens: number;
  raw_content_stored: 0 | 1;
  context_summary_json: string;
  token_breakdown_json: string;
  files_json: string;
  tool_schemas_json: string;
  mcp_servers_json: string;
  memories_json: string;
  sensitive_flags_json: string;
  repeated_regions_json: string;
  stale_items_json: string;
  trace_id: string | null;
  correlation_id: string | null;
};

const SAFE_KEYS = new Set([
  "id",
  "name",
  "path",
  "uri",
  "kind",
  "type",
  "source",
  "reason",
  "tokens",
  "estimatedTokens",
  "inputTokens",
  "bytes",
  "server",
  "tool",
  "ageTurns",
  "lastUsedAt",
  "verified",
  "verifierStatus",
  "status"
]);

export function recordContextPackage(host: SurfaceHost, input: ContextPackageInput): { contextPackageId: string; stored: boolean } {
  const contextPackageId = newId("ctxpkg");
  const files = sanitizedArray(input.contextPackage.files ?? input.contextPackage.fileRefs ?? input.contextPackage.contextFiles);
  const packageToolSchemas = input.toolSchemas.length > 0 ? input.toolSchemas : getArray(input.contextPackage, "toolSchemas");
  const packageMcpServers = input.mcpServers.length > 0 ? input.mcpServers : getArray(input.contextPackage, "mcpServers");
  const packageMemories = input.memories.length > 0 ? input.memories : getArray(input.contextPackage, "memories");
  const repeatedRegions = sanitizedArray(input.contextPackage.repeatedRegions ?? input.contextPackage.duplicates ?? input.contextPackage.repeatedContext);
  const staleItems = sanitizedArray(input.contextPackage.staleItems ?? input.contextPackage.staleContext);
  const tokenBreakdown = sanitizeTokenBreakdown(input.contextPackage.tokenBreakdown ?? input.contextSummary.tokenBreakdown ?? {});

  host.db.db
    .prepare(
      `INSERT INTO context_packages (
        id, preflight_request_id, session_id, agent_id, host_name, workspace_root,
        client, task_type, input_tokens, raw_content_stored, context_summary_json,
        token_breakdown_json, files_json, tool_schemas_json, mcp_servers_json,
        memories_json, sensitive_flags_json, repeated_regions_json, stale_items_json,
        origin_peer_id, trace_id, correlation_id, test_mode
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .run(
      contextPackageId,
      input.preflightId,
      input.sessionId,
      input.agentId,
      input.hostName,
      input.workspaceRoot,
      input.client,
      input.taskType,
      input.inputTokens,
      JSON.stringify(sanitizedRecord(input.contextSummary)),
      JSON.stringify(tokenBreakdown),
      JSON.stringify(files),
      JSON.stringify(sanitizedArray(packageToolSchemas)),
      JSON.stringify(sanitizedArray(packageMcpServers)),
      JSON.stringify(sanitizedArray(packageMemories)),
      JSON.stringify(input.sensitiveFlags),
      JSON.stringify(repeatedRegions),
      JSON.stringify(staleItems),
      input.originPeerId,
      input.traceId ?? null,
      input.correlationId ?? null,
      input.testMode ? 1 : 0
    );
  return { contextPackageId, stored: true };
}

export function fabricInspectContextPackage(host: SurfaceHost, input: unknown, context: CallContext): Record<string, unknown> {
  const session = host.requireSession(context);
  const requestId = getString(input, "requestId");
  const workspaceRoot = getOptionalString(input, "workspaceRoot") ?? context.workspaceRoot ?? session.workspace_root;
  const row = host.db.db
    .prepare("SELECT * FROM context_packages WHERE preflight_request_id = ? AND workspace_root = ?")
    .get(requestId, workspaceRoot) as ContextPackageRow | undefined;
  if (!row) {
    throw new FabricError("CONTEXT_PACKAGE_NOT_FOUND", `Context package not found for preflight: ${requestId}`, false);
  }

  const files = safeJsonArray(row.files_json);
  const toolSchemas = safeJsonArray(row.tool_schemas_json);
  const mcpServers = safeJsonArray(row.mcp_servers_json);
  const memories = safeJsonArray(row.memories_json);
  const sensitiveFlags = safeJsonArray(row.sensitive_flags_json).filter((item): item is string => typeof item === "string");
  const repeatedRegions = safeJsonArray(row.repeated_regions_json);
  const staleItems = safeJsonArray(row.stale_items_json);
  const tokenBreakdown = safeJsonRecord(row.token_breakdown_json);
  const warnings = contextWarnings({
    inputTokens: row.input_tokens,
    files,
    toolSchemas,
    mcpServers,
    memories,
    sensitiveFlags,
    repeatedRegions,
    staleItems,
    tokenBreakdown
  });
  const analysis = contextAnalysis({
    inputTokens: row.input_tokens,
    files,
    toolSchemas,
    mcpServers,
    memories,
    sensitiveFlags,
    repeatedRegions,
    staleItems,
    tokenBreakdown,
    warnings
  });

  return {
    requestId,
    contextPackageId: row.id,
    capturedAt: row.ts,
    workspaceRoot: row.workspace_root,
    client: row.client,
    taskType: row.task_type,
    rawContentStored: row.raw_content_stored === 1,
    summary: {
      inputTokens: row.input_tokens,
      fileCount: files.length,
      toolSchemaCount: toolSchemas.length,
      mcpServerCount: mcpServers.length,
      memoryCount: memories.length,
      sensitiveFlagCount: sensitiveFlags.length,
      repeatedRegionCount: repeatedRegions.length,
      staleItemCount: staleItems.length
    },
    tokenBreakdown,
    files,
    toolSchemas,
    mcpServers,
    memories,
    sensitiveFlags,
    repeatedRegions,
    staleItems,
    warnings,
    analysis
  };
}

function sanitizedArray(value: unknown): Record<string, unknown>[] {
  if (!Array.isArray(value)) return [];
  return value.map((item) => sanitizeItem(item)).filter((item) => Object.keys(item).length > 0);
}

function sanitizedRecord(value: unknown): Record<string, unknown> {
  const record = asRecord(value);
  const result: Record<string, unknown> = {};
  for (const [key, item] of Object.entries(record)) {
    if (SAFE_KEYS.has(key) || /tokens|count|bytes|size|reason|source|kind|type|status/i.test(key)) {
      result[key] = sanitizeScalar(item);
    }
  }
  return result;
}

function sanitizeTokenBreakdown(value: unknown): Record<string, unknown> {
  const record = asRecord(value);
  const result: Record<string, unknown> = {};
  for (const [key, item] of Object.entries(record)) {
    if (typeof item === "number" && Number.isFinite(item)) {
      result[key] = item;
    } else if (item && typeof item === "object" && !Array.isArray(item)) {
      const nested = sanitizeTokenBreakdown(item);
      if (Object.keys(nested).length > 0) result[key] = nested;
    }
  }
  return result;
}

function sanitizeItem(value: unknown): Record<string, unknown> {
  if (typeof value === "string") return { name: value };
  const record = asRecord(value);
  const result: Record<string, unknown> = {};
  for (const [key, item] of Object.entries(record)) {
    if (SAFE_KEYS.has(key) || /tokens|count|bytes|size|reason|source|kind|type|status/i.test(key)) {
      result[key] = sanitizeScalar(item);
    }
  }
  return result;
}

function sanitizeScalar(value: unknown): unknown {
  if (value === null || value === undefined) return value;
  if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") return value;
  if (Array.isArray(value)) return value.map(sanitizeScalar).filter((item) => item !== undefined);
  const record = asRecord(value);
  return Object.keys(record).length > 0 ? sanitizedRecord(record) : undefined;
}

function contextWarnings(input: {
  inputTokens: number;
  files: unknown[];
  toolSchemas: unknown[];
  mcpServers: unknown[];
  memories: unknown[];
  sensitiveFlags: string[];
  repeatedRegions: unknown[];
  staleItems: unknown[];
  tokenBreakdown: Record<string, unknown>;
}): string[] {
  const warnings: string[] = [];
  if (input.inputTokens >= 50_000) warnings.push(`Large context package (${input.inputTokens} tokens).`);
  if (input.toolSchemas.length > 5) warnings.push(`Large tool schema set (${input.toolSchemas.length}); consider dropping unused tools.`);
  if (input.sensitiveFlags.length > 0) warnings.push(`Sensitive flags present: ${input.sensitiveFlags.join(", ")}.`);
  if (input.repeatedRegions.length > 0) warnings.push(`Repeated context detected (${input.repeatedRegions.length} regions).`);
  if (input.staleItems.length > 0) warnings.push(`Stale context detected (${input.staleItems.length} items).`);
  if (input.memories.some((memory) => asRecord(memory).verified === false || asRecord(memory).verifierStatus === "unverified")) {
    warnings.push("Unverified memory is present in the context package.");
  }
  if (
    input.files.length === 0 &&
    input.toolSchemas.length === 0 &&
    input.mcpServers.length === 0 &&
    Object.keys(input.tokenBreakdown).length === 0
  ) {
    warnings.push("Client did not provide file/tool/MCP/token breakdown details.");
  }
  return warnings;
}

function contextAnalysis(input: {
  inputTokens: number;
  files: unknown[];
  toolSchemas: unknown[];
  mcpServers: unknown[];
  memories: unknown[];
  sensitiveFlags: string[];
  repeatedRegions: unknown[];
  staleItems: unknown[];
  tokenBreakdown: Record<string, unknown>;
  warnings: string[];
}): Record<string, unknown> {
  const repeatedTokens = sumTokenHints(input.repeatedRegions);
  const staleTokens = sumTokenHints(input.staleItems);
  const unverifiedMemories = input.memories.filter((memory) => asRecord(memory).verified === false || asRecord(memory).verifierStatus === "unverified");
  const knownBreakdownTokens = sumBreakdownTokens(input.tokenBreakdown);
  const estimatedWasteTokens = Math.min(
    input.inputTokens,
    repeatedTokens + staleTokens + sumTokenHints(unverifiedMemories) + Math.max(0, input.toolSchemas.length - 5) * 1_000
  );
  const severity =
    input.sensitiveFlags.length > 0 || input.inputTokens >= 200_000
      ? "blocker"
      : estimatedWasteTokens >= 20_000 || input.inputTokens >= 50_000 || input.toolSchemas.length > 8
        ? "high"
        : estimatedWasteTokens > 0 || input.toolSchemas.length > 5 || unverifiedMemories.length > 0
          ? "medium"
          : "low";
  return {
    severity,
    shouldCompactBeforeModel: severity === "blocker" || input.inputTokens >= 50_000 || input.sensitiveFlags.length > 0,
    estimatedWasteTokens,
    estimatedWasteRatio: input.inputTokens > 0 ? roundRatio(estimatedWasteTokens / input.inputTokens) : 0,
    knownBreakdownTokens,
    breakdownCoverage: input.inputTokens > 0 ? roundRatio(knownBreakdownTokens / input.inputTokens) : 0,
    largestFiles: largestByTokens(input.files, 8),
    largestToolSchemas: largestByTokens(input.toolSchemas, 8),
    largestMemories: largestByTokens(input.memories, 8),
    repeatedTokenEstimate: repeatedTokens,
    staleTokenEstimate: staleTokens,
    unverifiedMemoryCount: unverifiedMemories.length,
    suggestedActions: contextSuggestedActions(input, {
      severity,
      estimatedWasteTokens,
      repeatedTokens,
      staleTokens,
      unverifiedMemoryCount: unverifiedMemories.length
    })
  };
}

function contextSuggestedActions(
  input: {
    inputTokens: number;
    files: unknown[];
    toolSchemas: unknown[];
    mcpServers: unknown[];
    memories: unknown[];
    sensitiveFlags: string[];
    repeatedRegions: unknown[];
    staleItems: unknown[];
    warnings: string[];
  },
  analysis: { severity: string; estimatedWasteTokens: number; repeatedTokens: number; staleTokens: number; unverifiedMemoryCount: number }
): Array<Record<string, unknown>> {
  const actions: Array<Record<string, unknown>> = [];
  if (input.sensitiveFlags.length > 0) {
    actions.push({
      action: "remove_sensitive_context",
      priority: "blocker",
      reason: `Sensitive flags present: ${input.sensitiveFlags.join(", ")}`,
      expectedImpact: "Required before model routing."
    });
  }
  if (input.inputTokens >= 50_000) {
    actions.push({
      action: "compact_context",
      priority: analysis.severity === "blocker" ? "high" : analysis.severity,
      reason: `Context package is ${input.inputTokens} tokens.`,
      expectedImpact: "Lower cost and approval risk."
    });
  }
  if (input.toolSchemas.length > 5) {
    actions.push({
      action: "trim_tool_schemas",
      priority: input.toolSchemas.length > 8 ? "high" : "medium",
      reason: `${input.toolSchemas.length} tool schemas are attached.`,
      expectedImpact: "Expose only task-relevant tools/MCP servers."
    });
  }
  if (analysis.repeatedTokens > 0 || input.repeatedRegions.length > 0) {
    actions.push({
      action: "deduplicate_context",
      priority: "medium",
      reason: `${input.repeatedRegions.length} repeated region(s) detected.`,
      expectedImpact: "Remove repeated file or message regions."
    });
  }
  if (analysis.staleTokens > 0 || input.staleItems.length > 0) {
    actions.push({
      action: "drop_stale_context",
      priority: "medium",
      reason: `${input.staleItems.length} stale item(s) detected.`,
      expectedImpact: "Prefer current files, recent commands, and fresh checkpoints."
    });
  }
  if (analysis.unverifiedMemoryCount > 0) {
    actions.push({
      action: "review_memory",
      priority: "medium",
      reason: `${analysis.unverifiedMemoryCount} unverified memory item(s) attached.`,
      expectedImpact: "Avoid polluting the request with untrusted memory."
    });
  }
  if (actions.length === 0 && input.warnings.length === 0) {
    actions.push({
      action: "proceed",
      priority: "low",
      reason: "No obvious context waste was detected.",
      expectedImpact: "Proceed through the model gate."
    });
  }
  return actions;
}

function largestByTokens(values: unknown[], limit: number): Array<Record<string, unknown>> {
  return values
    .map((value) => {
      const record = asRecord(value);
      return {
        ...record,
        tokens: tokenHint(record)
      };
    })
    .filter((record) => record.tokens > 0)
    .sort((left, right) => Number(right.tokens) - Number(left.tokens))
    .slice(0, limit);
}

function sumTokenHints(values: unknown[]): number {
  return values.reduce<number>((sum, value) => sum + tokenHint(asRecord(value)), 0);
}

function tokenHint(record: Record<string, unknown>): number {
  for (const key of ["tokens", "estimatedTokens", "inputTokens"]) {
    const value = record[key];
    if (typeof value === "number" && Number.isFinite(value)) return Math.max(0, Math.round(value));
  }
  return 0;
}

function sumBreakdownTokens(value: unknown): number {
  if (typeof value === "number" && Number.isFinite(value)) return Math.max(0, Math.round(value));
  if (!value || typeof value !== "object" || Array.isArray(value)) return 0;
  return Object.values(value as Record<string, unknown>).reduce<number>((sum, item) => sum + sumBreakdownTokens(item), 0);
}

function roundRatio(value: number): number {
  return Math.round(value * 10_000) / 10_000;
}

File: workers/smolagents-worker/src/smolagents_worker/context_inspector.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class ContextFinding:
    severity: str
    title: str
    detail: str
    action: str


def load_context_inspection(path: str) -> dict[str, Any]:
    value = json.loads(Path(path).read_text(encoding="utf-8"))
    if not isinstance(value, dict):
        raise ValueError("context inspection JSON must be an object")
    return value


def analyze_context_package(inspection: dict[str, Any]) -> list[ContextFinding]:
    summary = _record(inspection.get("summary"))
    findings: list[ContextFinding] = []
    input_tokens = _int(summary.get("inputTokens"))
    file_count = _int(summary.get("fileCount"))
    tool_schema_count = _int(summary.get("toolSchemaCount"))
    mcp_server_count = _int(summary.get("mcpServerCount"))
    memory_count = _int(summary.get("memoryCount"))
    sensitive_flag_count = _int(summary.get("sensitiveFlagCount"))
    repeated_count = _int(summary.get("repeatedRegionCount"))
    stale_count = _int(summary.get("staleItemCount"))

    if input_tokens >= 200_000:
        findings.append(
            ContextFinding(
                "breakglass",
                "Context is in breakglass territory",
                f"The package is {input_tokens} input tokens before reserved output.",
                "Require explicit user approval, compact first, or move to a high-context/local route.",
            )
        )
    elif input_tokens >= 50_000:
        findings.append(
            ContextFinding(
                "high",
                "Context is large",
                f"The package is {input_tokens} input tokens.",
                "Inspect the largest files, stale items, and tool schemas before using a premium model.",
            )
        )
    elif input_tokens >= 20_000:
        findings.append(
            ContextFinding(
                "medium",
                "Context is moderately large",
                f"The package is {input_tokens} input tokens.",
                "Prefer a cheaper model or compact low-value logs/docs if the task is routine.",
            )
        )

    if tool_schema_count > 10:
        findings.append(
            ContextFinding(
                "high",
                "Too many tool schemas",
                f"{tool_schema_count} tool schemas are included.",
                "Drop unused MCP servers/tools before the model call.",
            )
        )
    elif tool_schema_count > 5:
        findings.append(
            ContextFinding(
                "medium",
                "Large tool schema set",
                f"{tool_schema_count} tool schemas are included.",
                "Keep only tools that are needed for this turn.",
            )
        )

    if mcp_server_count > 5:
        findings.append(
            ContextFinding(
                "medium",
                "Many MCP servers are in context",
                f"{mcp_server_count} MCP servers are represented.",
                "Disable or omit irrelevant servers for this call.",
            )
        )

    if sensitive_flag_count > 0:
        flags = ", ".join(str(item) for item in _list(inspection.get("sensitiveFlags")))
        findings.append(
            ContextFinding(
                "breakglass",
                "Sensitive context is present",
                flags or f"{sensitive_flag_count} sensitive flags are present.",
                "Require human approval and consider removing secrets, cookies, production data, or external-action context.",
            )
        )

    if repeated_count > 0:
        findings.append(
            ContextFinding(
                "medium",
                "Repeated context detected",
                f"{repeated_count} repeated regions were reported.",
                "Remove duplicated file/log/doc blocks before paying for a premium model.",
            )
        )

    if stale_count > 0:
        findings.append(
            ContextFinding(
                "medium",
                "Stale context detected",
                f"{stale_count} stale items were reported.",
                "Drop old logs, obsolete files, and context not used in recent turns.",
            )
        )

    unverified_memories = [item for item in _list(inspection.get("memories")) if _record(item).get("verified") is False or _record(item).get("verifierStatus") == "unverified"]
    if unverified_memories:
        findings.append(
            ContextFinding(
                "medium",
                "Unverified memories are present",
                f"{len(unverified_memories)} unverified memories are included out of {memory_count}.",
                "Use verified memories only for automatic injection; keep these as browseable context or ask for confirmation.",
            )
        )

    if file_count == 0 and tool_schema_count == 0 and mcp_server_count == 0 and not inspection.get("tokenBreakdown"):
        findings.append(
            ContextFinding(
                "high",
                "Context breakdown is missing",
                "The caller did not provide files, tools, MCP servers, or token breakdown metadata.",
                "Treat the preflight as low-confidence and improve client instrumentation before trusting routing decisions.",
            )
        )

    if not findings:
        findings.append(ContextFinding("low", "No obvious context waste detected", "The sanitized package did not trigger the default waste heuristics.", "Allow the call if the model route and budget are appropriate."))
    return findings


def render_context_report(inspection: dict[str, Any], findings: list[ContextFinding]) -> str:
    summary = _record(inspection.get("summary"))
    lines = [
        "# context inspection report",
        "",
        f"Request: `{inspection.get('requestId', 'unknown')}`",
        f"Context package: `{inspection.get('contextPackageId', 'unknown')}`",
        f"Workspace: `{inspection.get('workspaceRoot', 'unknown')}`",
        f"Client: `{inspection.get('client', 'unknown')}`",
        f"Task type: `{inspection.get('taskType', 'unknown')}`",
        f"Raw content stored: `{inspection.get('rawContentStored', False)}`",
        "",
        "## Summary",
        "",
    ]
    for key in (
        "inputTokens",
        "fileCount",
        "toolSchemaCount",
        "mcpServerCount",
        "memoryCount",
        "sensitiveFlagCount",
        "repeatedRegionCount",
        "staleItemCount",
    ):
        lines.append(f"- {key}: `{summary.get(key, 0)}`")

    lines.extend(["", "## Findings", ""])
    for finding in findings:
        lines.extend(
            [
                f"### {finding.severity.upper()}: {finding.title}",
                "",
                finding.detail,
                "",
                f"Action: {finding.action}",
                "",
            ]
        )

    lines.extend(["## Largest Files", ""])
    for item in _largest_items(_list(inspection.get("files"))):
        lines.append(_format_item(item))
    if len(lines) >= 2 and lines[-1] == "## Largest Files":
        lines.append("No file metadata supplied.")

    lines.extend(["", "## Tool Schemas", ""])
    for item in _largest_items(_list(inspection.get("toolSchemas"))):
        lines.append(_format_item(item))
    if lines[-1] == "## Tool Schemas":
        lines.append("No tool schema metadata supplied.")

    lines.extend(["", "## MCP Servers", ""])
    for item in _list(inspection.get("mcpServers")):
        lines.append(_format_item(_record(item)))
    if lines[-1] == "## MCP Servers":
        lines.append("No MCP server metadata supplied.")

    warnings = [str(item) for item in _list(inspection.get("warnings"))]
    lines.extend(["", "## Daemon Warnings", ""])
    if warnings:
        lines.extend(f"- {warning}" for warning in warnings)
    else:
        lines.append("No daemon warnings supplied.")
    lines.append("")
    return "\n".join(lines)


def write_context_report(output_dir: str, report: str, *, filename: str = "context-inspection-report.md") -> Path:
    output_path = Path(output_dir).resolve()
    output_path.mkdir(parents=True, exist_ok=True)
    target = output_path / filename
    target.write_text(report, encoding="utf-8")
    return target


def _largest_items(items: list[Any], limit: int = 10) -> list[dict[str, Any]]:
    records = [_record(item) for item in items]
    records.sort(key=lambda item: _item_size(item), reverse=True)
    return records[:limit]


def _item_size(item: dict[str, Any]) -> int:
    for key in ("tokens", "estimatedTokens", "inputTokens", "bytes", "size"):
        value = _int(item.get(key))
        if value:
            return value
    return 0


def _format_item(item: dict[str, Any]) -> str:
    name = item.get("path") or item.get("name") or item.get("server") or item.get("tool") or item.get("id") or "unknown"
    size = _item_size(item)
    reason = item.get("reason")
    suffix = f", reason={reason}" if reason else ""
    return f"- `{name}` ({size} tokens/bytes{suffix})"


def _record(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _list(value: Any) -> list[Any]:
    return value if isinstance(value, list) else []


def _int(value: Any) -> int:
    if isinstance(value, bool):
        return int(value)
    if isinstance(value, (int, float)):
        return int(value)
    return 0
